flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -19,432 +19,105 @@
 import org.apache.flink.runtime.taskmanager.TaskActions;
-import org.apache.flink.runtime.taskmanager.TaskManager;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
-import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicInteger;
 
-import static org.apache.flink.util.Preconditions.checkArgument;
-import static org.apache.flink.util.Preconditions.checkElementIndex;
 import static org.apache.flink.util.Preconditions.checkNotNull;
-import static org.apache.flink.util.Preconditions.checkState;
 
 /**
- * A result partition for data produced by a single task.
- *
- * <p>This class is the runtime part of a logical {@link IntermediateResultPartition}. Essentially,
- * a result partition is a collection of {@link Buffer} instances. The buffers are organized in one
- * or more {@link ResultSubpartition} instances, which further partition the data depending on the
- * number of consuming tasks and the data {@link DistributionPattern}.
- *
- * <p>Tasks, which consume a result partition have to request one of its subpartitions. The request
- * happens either remotely (see {@link RemoteInputChannel}) or locally (see {@link LocalInputChannel})
- *
- * <h2>Life-cycle</h2>
- *
- * <p>The life-cycle of each result partition has three (possibly overlapping) phases:
- * <ol>
- * <li><strong>Produce</strong>: </li>
- * <li><strong>Consume</strong>: </li>
- * <li><strong>Release</strong>: </li>
- * </ol>
- *
- * <h2>Lazy deployment and updates of consuming tasks</h2>
- *
- * <p>Before a consuming task can request the result, it has to be deployed. The time of deployment
- * depends on the PIPELINED vs. BLOCKING characteristic of the result partition. With pipelined
- * results, receivers are deployed as soon as the first buffer is added to the result partition.
- * With blocking results on the other hand, receivers are deployed after the partition is finished.
- *
- * <h2>Buffer management</h2>
- *
- * <h2>State management</h2>
+ * A wrapper of result partition writer for handling notification of the consumable
+ * partition which is added a {@link BufferConsumer} or finished.
  */
-public class ResultPartition implements ResultPartitionWriter, BufferPoolOwner {
-
-	private static final Logger LOG = LoggerFactory.getLogger(ResultPartition.class);
-
-	private final String owningTaskName;
+public class ResultPartition implements ResultPartitionWriter {
 
 	private final TaskActions taskActions;
 
 	private final JobID jobId;
 
-	private final ResultPartitionID partitionId;
-
-	/** Type of this partition. Defines the concrete subpartition implementation to use. */
 	private final ResultPartitionType partitionType;
 
-	/** The subpartitions of this partition. At least one. */
-	private final ResultSubpartition[] subpartitions;
-
-	private final ResultPartitionManager partitionManager;
+	private final ResultPartitionWriter partitionWriter;
```

Comment:
At the moment, up to this PR, `ResultPartitionWriter` interface is not abstracted fully yet.
So we see this PR as just one step closer to the refactoring of `ResultPartitionWriter` interface.

I would keep this field explicitly as `NetworkResultPartition` and keep a getter method for it.
Then we can keep `ResultPartition` where possible outside of `NetworkEnviroment` and whenever we need to call a specific method of `NetworkResultPartition` (not `ResultPartitionWriter`), 
we can use `resultPartition.getNetworkResultPartition().specificMethod()`.

Later, we can quickly drop `getNetworkResultPartition()` method and see want we miss to switch from `ResultPartition` outside of `NetworkEnviroment` to `ResultPartitionWriter`.

This way we could have less changes in this PR. Especially, we do not need a `NetworkResultPartition` array in Task.

What do you think?

Reply:
In the final form, `ResultPartition` should not implement `ResultPartitionWriter` and it could be put in the `TaskManager` package instead of current network package. And the field in `ResultPartition` should be regular `ResultPartitionWriter` finally.

I could understand your concern. If we keep special `NetworkResultPartition` in `ResultPartition` temporarily, it would bring other troubles after I have a try. The mainly involved ones are `Environment` and `RecordWriter`.

- `RecordWriter` -> `ResultPartition` -> `NetworkResultPartition`. There are many other implementations for `ResultPartitionWriter` used for tests. In `RecordWriterTest` we could not construct the proper `RecordWriter` based on `ResultPartition` which contains specific `NetworkResultPartition`.

- `Environment` should keep `ResultPartition` instead of `ResultPartitionWriter`, because it would be used for constructing `RecordWriter` later. If we keep specific `NetworkResultPartition` in `ResultPartition`, then it could not return proper `ResultPartition` for other `Environment` implementations, such as `StreamMockEnvironment`, `MockEnvironment` etc, because they maintain other `ResultPartitionWriter` implementations.

So I think the current way is more easy to handle, because we only need refactor `ResultPartition` and `Task` finally. Another option is we suspend this PR currently, you could firstly review other PRs which are involved in `ResultPartitionWriter` interface refactor. After we make a final interface form, then we could come back to maintain `ResultPartitionWriter` in `ResultPartition` finally. What do you think?

Reply:
I checked the other related PRs. The final form of `ResultPartitionWriter` interface would only cover some basic methods of `getBufferProvider`, `addBufferConsumer`, `flushAll`, `flush` after merging #7822 to rebase. Then the constructor of `ResultPartition` would cover some other fields such as `numberOfSubpartitions`, `numberOfTargetKeyGroups` which could be got directly from `ResultPartitionDeploymentDescriptor`.

Reply:
Why should `ResultPartition` (notification) not implement eventually `ResultPartitionWriter`?
Then everything could depend just on `ResultPartitionWriter` interface.
The `ResultPartition` could become:
`ResultPartitionWriterWithConsumableNotification implements ResultPartitionWriter` in Task package.
Task could just wrap `ShuffleService.ResultPartitionWriter` (`NetworkResultPartition` for netty) with `ResultPartitionWriterWithConsumableNotification` and pass it to `RecordWriter` also as `ResultPartitionWriter`.
Also tests could do the same if notification is relevant for them or use just `NetworkResultPartition` as `ResultPartitionWriter`.

but I think I see your concern at the moment, as we do not keep `NetworkResultPartition` per Task in `NetworkEnviroment` yet, probably having 2 arrays for writers is the simplest at this step.
We can also work on other PRs and see how it goes, maybe it will simplify this one.

Reply:
It is no problem to let `ResultPartition` also implement `ResultPartitionWriter`. And it might also make sense to not implement the interface. Because `ResultPartition` would be refactored to task package finally, if it implements the interface, it seems still coupled with network interface. In addition, there already exists a `ResultPartitionWriter` component in `ResultPartition`. If the `ResultPartition` itself is also a `ResultPartitionWriter`, it might seem a bit duplicated.

ShuffleService would create regular `ResultPartitionWriter` which is wrapped  into `ResultPartitionWithConsumableNotification` in task stack. And ` ResultPartitionWithConsumableNotification` would be passed into `RuntimeEnvironment` to be used for creating `RecordWriter` later. It might also makes sense to reference with `ResultPartitionWithConsumableNotification` instead of `ResultPartitionWriter` in `RecordWriter` and `RuntimeEnvironment`. 

As you said, we could keep the current status. And after other parts ready, we could further check how it goes. :)